### PR TITLE
Collect connection pool statistics periodically

### LINKED ISSUE #150: Add monitoring of connection pools
It'd be nice to get some data on how application connection pools are doing. This should be doable for things like thrift_pool that we control completely and sqlalchemy that give us introspection events.

### PR DESCRIPTION
**Pull request by spladug**
This adds a new ability to ContextFactory objects that hooks them into
the server runtime metrics system and asks them to periodically produce
metrics.

The SQLAlchemy and Thrift context clients are now instrumented to
periodically report their connection pool statistics to give more
insight into what's going on in there.

For some simple context clients like this:

```python
baseplate.add_to_context(
    "downstream", ThriftContextFactory(thrift_pool, BaseplateService.Client))
baseplate.add_to_context("db", SQLAlchemyEngineContextFactory(engine))
```

The following metrics might be sent:

```
example_server.runtime.hostname.PID13814.clients.downstream.size:10|g
example_server.runtime.hostname.PID13814.clients.downstream.checkedout:0|g
example_server.runtime.hostname.PID13814.clients.db.size:5|g
example_server.runtime.hostname.PID13814.clients.db.checkedin:0|g
example_server.runtime.hostname.PID13814.clients.db.overflow:0|g
example_server.runtime.hostname.PID13814.clients.db.checkedout:0|g
```

This starts to fix #150.

### CODE REVIEW

**Comment by alienth** [comment]
Related thought I had while looking at this: I wonder if it'd be possible to have a gauge metric which increments whenever we try to make a query and the pool is exhausted?

File: baseplate/server/runtime_monitor.py [hunk 1]
```diff
@@ -108,7 +124,8 @@ def _report_runtime_metrics_periodically(metrics_client, reporters):
 
 
 def start(server_config, application, pool):
-    if not hasattr(application, "baseplate") or not application.baseplate._metrics_client:
+    baseplate = getattr(application, "baseplate", None)
+    if not baseplate or not baseplate._metrics_client:
```
Comment (alienth) [inline baseplate/server/runtime_monitor.py]: Unrelated change, or am I missing something?

Comment (spladug) [reply baseplate/server/runtime_monitor.py]: I wanted to keep the `baseplate` object around so I could use it when building the reporter thing down below. Before it was just ephemerally present in this check to get the metrics off of it.

**Comment by spladug** [comment]
> Related thought I had while looking at this, doesn't need to be solved here: I wonder if it'd be possible to have a counter metric which increments whenever we try to make a query and the pool is exhausted?

Yeah, I was thinking of putting a timer around the block for an empty slot but currently the pool is "below" that level and has no concept of request or trace context so it'd be hard to get at that. Will need to play with some reworkings for that to make sense.

File: baseplate/context/sqlalchemy.py [hunk 2]
```diff
@@ -79,6 +80,16 @@ def __init__(self, engine):
         event.listen(self.engine, "after_cursor_execute", self.on_after_execute)
         event.listen(self.engine, "dbapi_error", self.on_dbapi_error)
 
+    def report_runtime_metrics(self, batch):
```
Comment (pacejackson) [inline baseplate/context/sqlalchemy.py]: Can we add a docstring to each of these to document the different metrics it reports?

File: baseplate/server/runtime_monitor.py [hunk 3]
```diff
@@ -81,6 +81,22 @@ def report(self, batch):
...
 
+class _BaseplateReporter:
+    def __init__(self, reporters):
+        self.reporters = reporters
+
+    def report(self, batch):
+        for name, reporter in self.reporters.items():
+            original_namespace = batch.namespace
+            try:
+                batch.namespace = b".".join((batch.namespace, b"clients", name.encode()))
```
[the remark this answers is not in the document]
Comment (spladug) [reply baseplate/server/runtime_monitor.py]: These are already separate from endpoint metrics because they're in the PID+hostname namespace.

File: baseplate/context/thrift.py [hunk 4]
```diff
@@ -48,6 +48,13 @@ def __init__(self, pool, client_cls):
             },
         )
 
+    def report_runtime_metrics(self, batch):
+        batch.gauge("size").replace(self.pool.size)
+        batch.gauge("checkedout").replace(self.pool.checkedout)
+        # it's hard to report "checkedin" currently because we can't
+        # distinguish easily between available connection slots that aren't
+        # instantiated and ones that have actual open connections.
```
[the remark this answers is not in the document]
Comment (spladug) [reply baseplate/context/thrift.py]: Yeah! I'm totally down for a different set of names. I definitely just copied sqlalchemy's for the thrift pool. Got any thoughts on what would be clearest? Do we have any prior art elsewhere?

[the remark this answers is not in the document]
Comment (pacejackson) [reply baseplate/context/thrift.py]: `size`: A+, would use again
`overflow`: I'm not 100% sure what this is but I would guess it's requests that are waiting for an available slot in the pool?  `overflow` seems appropriate or `waiting` could also work.
`checkedin`: `available`, `open`.
`checkedout`: `in_use`/`inuse`/`in-use`, `active`, `used`

I don't think that any of the sqlalchemy names are crazy or super specific to the library so I think they work too.

[the remark this answers is not in the document]
Comment (pacejackson) [reply baseplate/context/thrift.py]: Note, I'm not _aware_ of any prior art.

[the remark this answers is not in the document]
Comment (pacejackson) [reply baseplate/context/thrift.py]: 🔕 Sort of related, should these be `pool.${metric}`?  That way we can potentially add other metrics that aren't explicitly related to the connection pools and have those name-spaced appropriately?  Or is that overkill?

[the remark this answers is not in the document]
Comment (spladug) [reply baseplate/context/thrift.py]: `overflow` is a SQLAlchemy pool thing. It's a number of connections that can be opened beyond the max size of the pool, but if they're returned to the pool they're closed rather than kept around. So they allow spike capacity but don't stick around for the long haul.

I'm cool with `available` and `in_use`. That sounds clear to me.

Yeah I think `pool.` prefix is fine since it'd allow easy wildcards.

[the remark this answers is not in the document]
Comment (pacejackson) [reply baseplate/context/thrift.py]: I like it all!

[the remark this answers is not in the document]
Comment (spladug) [reply baseplate/context/thrift.py]: I had to change to `open_and_available` because it's different from `size - in_use` in that it's explicitly telling us how many TCP connections are open but checked back into the pool; it does not count how many connection slots are free.

**Comment by spladug** [comment]
:haircut_man:

**Comment by pacejackson** [comment]
One thought I had, would there be issues with this sending too many metrics or having too high a path cardinality given the `${hostname}.PID${pid}` naming scheme?

CC: @ckwang8128

**Comment by spladug** [comment]
That concern already exists for the other runtime metrics. We're doing aggregation in telegraf before it gets to the stats backend.

**Comment by pacejackson** [comment]
Ah, I wasn't aware of those, great! 🐟
